ogre2/src/Ogre2SegmentationCamera.cc

```diff
@@ -0,0 +1,647 @@
+  /// with colored version for segmentation
+  public: SegmentationMaterialSwitcher *materialSwitcher {nullptr};
+
+  /// \brief Compositor Manager to create workspace
+  public: Ogre::CompositorManager2 *ogreCompositorManager {nullptr};
+
+  /// \brief Workspace to interface with render texture
+  public: Ogre::CompositorWorkspace *ogreCompositorWorkspace {nullptr};
+
+  /// \brief Workspace Definition
+  public: std::string workspaceDefinition;
+
+  /// \brief Render Texture to store the final segmentation data
+  public: Ogre::RenderTexture *ogreRenderTexture {nullptr};
+
+  /// \brief Texture to create the render texture from.
+  public: Ogre::TexturePtr ogreTexture;
+
+  /// \brief Pixel Box to copy render texture data to a buffer
+  public: Ogre::PixelBox *pixelBox {nullptr};
+
+  /// \brief buffer to store render texture data & to be sent to listeners
+  public: uint8_t *buffer = nullptr;
+
+  /// \brief dummy render texture to set image dims
+  public: Ogre2RenderTexturePtr dummyTexture {nullptr};
+
+  /// \brief New Segmentation Frame Event to notify listeners with new data
+  public: ignition::common::EventT<void(const uint8_t *, unsigned int,
+        unsigned int, unsigned int, const std::string &)>
+        newSegmentationFrame;
+
+  /// \brief Image / Render Texture Format
+  public: Ogre::PixelFormat format = Ogre::PF_R8G8B8;
+};
+
+/////////////////////////////////////////////////
+Ogre2SegmentationCamera::Ogre2SegmentationCamera() :
+  dataPtr(new Ogre2SegmentationCameraPrivate())
+{
+}
+
+/////////////////////////////////////////////////
+Ogre2SegmentationCamera::~Ogre2SegmentationCamera()
+{
+  this->Destroy();
+}
+
+/////////////////////////////////////////////////
+void Ogre2SegmentationCamera::Init()
+{
+  BaseCamera::Init();
+  this->CreateCamera();
+  this->CreateRenderTexture();
+
+  this->dataPtr->materialSwitcher =
+    new SegmentationMaterialSwitcher(this->scene);
+}
+
+/////////////////////////////////////////////////
+void Ogre2SegmentationCamera::CreateCamera()
+{
+  auto ogreScene = this->scene->OgreSceneManager();
+  if (ogreScene == nullptr)
+  {
+    ignerr << "Scene manager cannot be obtained" << std::endl;
+    return;
+  }
+
+  this->ogreCamera = ogreScene->createCamera(this->Name());
+  if (this->ogreCamera == nullptr)
+  {
+    ignerr << "Ogre camera cannot be created" << std::endl;
+    return;
+  }
+
+  this->ogreCamera->detachFromParent();
+  this->ogreNode->attachObject(this->ogreCamera);
+
+  // rotate to ignition gazebo coord.
+  this->ogreCamera->yaw(Ogre::Degree(-90));
+  this->ogreCamera->roll(Ogre::Degree(-90));
+  this->ogreCamera->setFixedYawAxis(false);
+
+  this->ogreCamera->setAutoAspectRatio(true);
+  this->ogreCamera->setRenderingDistance(100);
+  this->ogreCamera->setProjectionType(Ogre::ProjectionType::PT_PERSPECTIVE);
+  this->ogreCamera->setCustomProjectionMatrix(false);
+}
+
+/////////////////////////////////////////////////
+void Ogre2SegmentationCamera::Destroy()
```

Comment:
Make sure you are deleting all of the raw pointers in the `Ogre2SegmentationPrivate` class (`dataPtr` member variable) here. Or, if you use `std::unique_ptr` instead of raw pointers as I suggested in my other comment, you don't need to worry about this :slightly_smiling_face: